graphula-core/src/Graphula.hs

```diff
@@ -445,69 +454,124 @@ type GraphulaNode m a =
+  }
+  deriving (Generic)
 
-  > nodeEdit @Dog (ownerId, veterinarianId) $ \dog ->
-  >   dog {name = "fido"}
--}
-nodeEditWith
-  :: forall a m
-   . GraphulaContext m '[a]
-  => Dependencies a
-  -> (a -> a)
-  -> m (Entity a)
-nodeEditWith dependencies edits = 10 `attemptsToInsertWith` do
-  x <- (`dependsOn` dependencies) . edits <$> generateNode
-  logNode x
-  pure x
+instance Semigroup (NodeOptions a) where
+  (<>) = gmappend
 
-{-|
-  Generate a value with data dependencies. This leverages 'HasDependencies' to
-  insert the specified data in the generated value.
+instance Monoid (NodeOptions a) where
+  mempty = gmempty
 
-  > nodeEdit @Dog (ownerId, veterinarianId)
--}
-nodeWith
-  :: forall a m . GraphulaContext m '[a] => Dependencies a -> m (Entity a)
-nodeWith = flip nodeEditWith id
+-- | Modify the node after it's been generated
+--
+-- @
+-- a <- root @A $ edit $ \a -> a { someField = True }
+-- @
+--
+edit :: (a -> a) -> NodeOptions a
+edit f = mempty
+  { nodeOptionsEdit = Endo $ fmap f
+  }
+
+-- | Require a node to satisfy the specified predicate
+--
+-- @
+-- a <- root @A $ suchThat $ (== True) . someField
+-- @
+--
+suchThat :: (a -> Bool) -> NodeOptions a
+suchThat f = mempty
+  { nodeOptionsEdit = Endo $ \ma -> do
+      a <- ma
+      a <$ guard (f a)
+  }
+
+-- | Override any automatic key generation with the specified key
+--
+-- @
+-- a <- root @A $ primaryKey $ AKey 1
+-- @
+--
+primaryKey :: Key a -> NodeOptions a
+primaryKey key = mempty
+  { nodeOptionsPrimaryKey = Last $ Just $ pure $ Just key
+  }
+
+-- | Generate a value that does not have any dependencies
+--
+-- @
+-- root @Dog mempty
+-- @
+--
+root
+  :: forall a m . (GraphulaContext m '[a], Dependencies a ~ ()) => NodeOptions a -> m (Entity a)
+root = node ()
 
 {-|
-  Generate and edit a value that does not have any dependencies.
+  Generate a value with data dependencies. This leverages
+  'HasDependencies' to insert the specified data in the generated value. All
+  dependency data is inserted after any editing operations.
 
-  > nodeEdit @Dog $ \dog -> dog {name = "fido"}
+  > node @Dog (ownerId, veterinarianId) mempty
+  > node @Dog (ownerId, vererinarianId) $ primaryKey $ DogKey 1
+  > node @Dog (ownerId, vererinarianId) $ edit $ \dog ->
+  >   dog {name = "fido"}
 -}
-nodeEdit
+node
   :: forall a m
-   . (GraphulaContext m '[a], Dependencies a ~ ())
-  => (a -> a)
+   . GraphulaContext m '[a]
+  => Dependencies a
+  -> NodeOptions a
   -> m (Entity a)
-nodeEdit = nodeEditWith ()
-
--- | Generate a value that does not have any dependencies
---
--- > node @Dog
-node
-  :: forall a m . (GraphulaContext m '[a], Dependencies a ~ ()) => m (Entity a)
-node = nodeWith ()
-
-attemptsToInsertWith
+node dependencies NodeOptions{..} = attempt 100 10 $ do
+  initial <- generateNode
+  for (appEndo nodeOptionsEdit $ Just initial) $ \edited -> do
+    let hydrated = edited `dependsOn` dependencies
+    logNode hydrated
```

Comment:
Is there an issue with `logNode` being here? If a node fails to insert, will `logNode` put multiple items in the log?

Reply:
This is how it was working before. I wasn't sure if I should change it. Also, if you do fail to insert, you probably want to have logged something.

Reply:
> This is how it was working before. 

Yup, but it might have been broken then too.

> Also, if you do fail to insert, you probably want to have logged something.

The log is for replay and inspection of the graph when a test fails. So you don't want data in there that isn't used.